### DISCUSSION TITLE
Getting category suggestions for a listing item using the Etsy openApi

### DISCUSSION THREAD

**Discussion by tahirkahnicloud** [etsy/open-api#990] [Q&A]
Hey guys I am not sure if this is available in v3 of the Etsy open API. I did not find a way to get category suggestions. Not sure but I think this existed in v2 and I do not see it in V3. Other than creating custom mapping algo in code, is there a simpler way of getting cat suggestions?

Thanks!

**Comment by tahirkahnicloud** [comment]
bump

**Comment by duongvd2106** [comment]
We had to do it manually.

1. In the first step, we get the entire Seller Taxonomy as a tree node.
2. We tried to flatten that tree node into a list of Taxonomies to get the full path of each node.
3. We implemented the search algorithm on that full path list.

Here is the application screen capture that we made.
I hope it helps you.

![Capture](https://github.com/etsy/open-api/assets/13653172/88d95750-c14b-4232-a83d-f72b3aac14ba)

**Reply by adiljaved76** [reply]
did you create a smart search dropdown combo box to list flatten tree? Is it possible to get some code sample shared? I'm trying to implement similar thing but need to show it in my data grid cell (a comboboxcell)

**Reply by duongvd2106** [reply]
Yes, that's right. I created this search input with autocomplete datasource taken from flattened tree.
Below is my tree flattening source code, written in Java language.
Hope it helps with your issue.

Taxonomy Object Wrapper
```java
//... 
import java.util.ArrayList;
import com.google.gson.annotations.SerializedName;

public class EtsyTaxonomyV3 {
	@SerializedName("id")
	private String taxonomyId;
	@SerializedName("level")
	private int level;
	@SerializedName("name")
	private String name;
	@SerializedName("parent_id")
	private String parentTaxonomyId;
	@SerializedName("children")
	private ArrayList<EtsyTaxonomyV3> arrChildNodes;
	@SerializedName("full_path_taxonomy_ids")
	private ArrayList<Integer> fullPathTaxonomyIds; // listed in order from root to leaf.
	
	private boolean isSellerTaxonomy = false; // calculated properties
	private String _str_full_path_taxonomy; // calculated properties

	public EtsyTaxonomyV3() {
	}
	
	public EtsyTaxonomyV3(boolean isSellerTaxonomy, String taxonomyId, int level, String name, String parentTaxonomyId, ArrayList<Integer> fullPathTaxonomyIds) {
		super();
		this.isSellerTaxonomy = isSellerTaxonomy;
		this.taxonomyId = taxonomyId;
		this.level = level;
		this.name = name;
		this.parentTaxonomyId = parentTaxonomyId;
		this.fullPathTaxonomyIds = fullPathTaxonomyIds;
	}

	public ArrayList<EtsyTaxonomyV3> getFlattenArrayListOfAllNodes(ArrayList<EtsyTaxonomyV3> arr){
		arr.add(this);
		for (EtsyTaxonomyV3 etsyTaxonomyV3 : getArrChildNodes()) {
			arr = etsyTaxonomyV3.getFlattenArrayListOfAllNodes(arr);
		}
		return arr;
	}
	public String calculateFullPathTaxonomyNamesAsString(ArrayList<EtsyTaxonomyV3> arrTaxonomies) {
		if (this.fullPathTaxonomyIds == null || this.fullPathTaxonomyIds.size() == 0) {
			return "";
		}
		String str = "";
		for (int i = 0; i < this.fullPathTaxonomyIds.size(); i++) {
			Integer parentTaxonomyId = this.fullPathTaxonomyIds.get(i);
			EtsyTaxonomyV3 parentTaxonomy = arrTaxonomies.stream().filter(taxonomy -> parentTaxonomyId.toString().equals(taxonomy.getTaxonomyId())).findFirst().orElse(null);
			if(parentTaxonomy != null) {
				str += ((i < this.fullPathTaxonomyIds.size() - 1) ? (parentTaxonomy.getName() + " > ") : parentTaxonomy.getName());
			}
		}
		return str;
	}
        //... getter & setter
}
``` 
After call GET request to "https://openapi.etsy.com/v3/application/seller-taxonomy/nodes", parse response Json to Object Wrapper
```java
//...
ArrayList<EtsyTaxonomyV3> arrEtsyTaxonomies = new ArrayList<EtsyTaxonomyV3>();
String serviceUrl = isSellerTaxonomy ? "seller-taxonomy/nodes" : "buyer-taxonomy/nodes";
try {
	String syncEtsyTaxonomyResults = this.getService(serviceUrl, true);
	if (!CommonUtil.isEmptyString(syncEtsyTaxonomyResults)) {
		ObjectMapper mapper = new ObjectMapper();
		JsonNode readTree = mapper.readTree(syncEtsyTaxonomyResults);
		int cnt = readTree.get("count").asInt();
		if (cnt > 0) {
			JsonNode taxonomiesTree = readTree.get("results");
			Iterator<JsonNode> iter = taxonomiesTree.iterator();
			while (iter.hasNext()) {
				JsonNode modules = iter.next();
				arrEtsyTaxonomies.add((EtsyTaxonomyV3) GsonUtil.convertJsonToObject(modules.toPrettyString(), EtsyTaxonomyV3.class));
			}
		}
		//...
	} 
        //...
} catch (Exception e) {
	e.printStackTrace();
	//...
}
//...
``` 
Then flatten the tree node array list to normal array list
```java
//...
// flatten tree node
ArrayList<EtsyTaxonomyV3> arrEtsyTaxonomiesSellerFlatted = new ArrayList<EtsyTaxonomyV3>();
for (EtsyTaxonomyV3 etsyTaxonomyV3: arrEtsyTaxonomies) {
	arrEtsyTaxonomiesSellerFlatted = etsyTaxonomyV3.getFlattenArrayListOfAllNodes(arrEtsyTaxonomiesSellerFlatted);
}
// construct full path taxonomy as text string for search & display
for (EtsyTaxonomyV3 etsyTaxonomyV3: arrEtsyTaxonomiesSellerFlatted) {
	etsyTaxonomyV3.set_str_full_path_taxonomy(etsyTaxonomyV3.calculateFullPathTaxonomyNamesAsString(arrEtsyTaxonomiesSellerFlatted));
}
//...
```

**Comment by tahirkahnicloud** [comment]
Are you page scraping? Also limiting the taxonomy to the seller's taxonomy implies that this would break if the seller tries to go beyond their taxonomy and attempt to list an item, correct?

**Reply by duongvd2106** [reply]
We use API V3 to get the full seller taxonomy node tree data.
You can refer to it here: https://developer.etsy.com/documentation/reference#tag/SellerTaxonomy
Once you get it, flatten the tree into a list.
Our End Users are only allowed to select the categories that we offer, each of our category will be mapped to a Seller Taxonomy Node in the tree.

Page scraping is strictly prohibited by Etsy. Don't do that way.

**Comment by tahirkahnicloud** [comment]
Unfortunately, our use case is pretty expansive. We are category matching between one platform to the other, so it is not as simple as this. I am not sure whether the GetCategorySuggestions existed in Etsy API V2 but based on what I have read it did exist. This would be super helpful, if not then developers like us will have to keep re-inventing the wheel with custom implementations.

**Comment by tahirkahnicloud** [comment]
BTW, we are already doing this but in c#. Our system populates the cats and subcats once and persists the response into different DB tables. That is not the issue. The issue we have is a classic last mile problem. Corelating categories/subcategories between two different platform with totally different taxonomies. Where Etsy could help is migrating the category suggestions to v3.
